src/rebar_prv_eunit.erl

```diff
@@ -22,507 +20,126 @@
-                                                               {module, ?MODULE},
-                                                               {bare, false},
-                                                               {deps, ?DEPS},
-                                                               {example, "rebar eunit"},
-                                                               {short_desc, "Run eunit tests on project apps."},
-                                                               {desc, ""},
-                                                               {opts, []}])),
-
+    Provider = providers:create([{name, ?PROVIDER},
+                                 {module, ?MODULE},
+                                 {deps, ?DEPS},
+                                 {bare, false},
+                                 {example, "rebar eunit"},
+                                 {short_desc, "Run EUnit Tests."},
+                                 {desc, ""},
+                                 {opts, eunit_opts(State)},
+                                 {profile, test}]),
+    State1 = rebar_state:add_provider(State, Provider),
     {ok, State1}.
 
 -spec do(rebar_state:t()) -> {ok, rebar_state:t()} | {error, string()}.
 do(State) ->
-    ok = ensure_dirs(),
-    %% Save code path
-    CodePath = setup_code_path(),
-    CompileOnly = rebar_state:get(State, compile_only, false),
-    {ok, SrcErls} = rebar_erlc_compiler:test_compile(State, "eunit",
-                                                     ?EUNIT_DIR),
-    case CompileOnly of
-        "true" ->
-            true = rebar_utils:cleanup_code_path(CodePath),
-            ?CONSOLE("Compiled modules for eunit~n", []);
-        false ->
-            run_eunit(State, CodePath, SrcErls)
+    ?INFO("Performing EUnit tests...", []),
+    {RawOpts, _} = rebar_state:command_parsed_args(State),
+    Opts = transform_opts(RawOpts, State),
+    ProjectApps = rebar_state:project_apps(State),
+    OutDir = case proplists:get_value(outdir, Opts, undefined) of
+        undefined -> filename:join([rebar_state:dir(State),
+                     ec_file:insecure_mkdtemp()]);
+        Path -> Path
     end,
-    {ok, State}.
+    ?DEBUG("Compiling EUnit instrumented modules in: ~p", [OutDir]),
+    lists:foreach(fun(App) ->
+                      AppDir = rebar_app_info:dir(App),
+                      C = rebar_config:consult(AppDir),
+                      S = rebar_state:new(State, C, AppDir),
+                      %% combine `erl_first_files` and `eunit_first_files` and adjust
+                      %% compile opts to include `eunit_compile_opts`, `{d, 'TEST'}`
+                      %% and `{src_dirs, "test"}`
+                      TestState = first_files(test_opts(S, OutDir)),
+                      ok = rebar_erlc_compiler:compile(TestState, AppDir)
+                  end, ProjectApps),
+    true = code:add_patha(OutDir),
```

Comment:
We should set the code_path back to what it was before at the end of this provider.